### PR TITLE
[12.0] [MIG] donation_base

### PR DESCRIPTION
**Pull request by nikul-serpentcs**
Donation Base
=============

This is the base module for donations. This module doesn't do anything in itself ;
it just adds some properties on products and partners and adds the *donation.tax.receipt* object.

### CODE REVIEW

File: donation_base/wizard/tax_receipt_print.py
Comment (tbaden) [inline donation_base/wizard/tax_receipt_print.py:32]: is this still needed or is this some historical code? @alexis-via

**Review by tbaden — APPROVED** [review]
LGTM, did a code review.
Just a question for alexis later on.

File: donation_base/wizard/tax_receipt_annual_create.py
Comment (tbaden) [inline donation_base/wizard/tax_receipt_annual_create.py]: on functional review we found this issue:
datetime is no longer used. Instead it works with fields.Date. This need fixing

If you handle dates and datetimes, you might need to adapt your code now that Odoo returns always native Python objects, so no more fields.Date/Datetime.from_string is needed. For converting old style date/datetime strings to Python object, use fields.Date.to_date and fields.Datetime.to_datetime methods.

File: donation_base/wizard/tax_receipt_annual_create.py
Comment (tbaden) [inline donation_base/wizard/tax_receipt_annual_create.py]: same here

**Review by tbaden — CHANGES_REQUESTED** [review]
we did some functional testing and found this one

@benbrich had an error on "create" of annual receipts for datetime

**Comment by alexis-via** [comment]
Great ! Could you up-port that commit we made yesterday on v11 ? https://github.com/OCA/donation/commit/070a78823b94eee01ca9a502e21929be84c77668

**Comment by nikul-serpentcs** [comment]
@alexis-via @tbaden @benbrich Code updated

**Comment by nikul-serpentcs** [comment]
In the above commit, Added a `donor` boolean on res.partner. https://github.com/OCA/donation/issues/43#issuecomment-536611002 point No. 2

**Comment by tbaden** [comment]
@nikul-serpentcs unfortunately travis is red
the test test_donation_check_type is now obsolete as products of type consu are now allowed.
please remove it.

**Comment by nikul-serpentcs** [comment]
@pedrobaeza @alexis-via  Can you please help here, Why Runbot is Fail, Any Error?

**Review by tbaden — APPROVED** [review]

File: donation_base/models/partner.py
Comment (jarroyomorales) [inline donation_base/models/partner.py:40]: Why would it throw an exception?

File: donation_base/models/product.py
Comment (jarroyomorales) [inline donation_base/models/product.py:69]: Not necessary

File: donation_base/models/product.py
Comment (jarroyomorales) [inline donation_base/models/product.py:76]: Not necessary

File: donation_base/models/donation_tax_receipt.py
Comment (jarroyomorales) [inline donation_base/models/donation_tax_receipt.py:62]: Can you do it the way it is usually done now? With the default "/"

File: donation_base/models/donation_tax_receipt.py
Comment (jarroyomorales) [inline donation_base/models/donation_tax_receipt.py:26]: Shouldnt it be computed?

File: donation_base/report/report_donationtax.xml
Comment (jarroyomorales) [inline donation_base/report/report_donationtax.xml:13]: Should it also display the lines?

**Review by jarroyomorales — COMMENTED** [review]
A couple of comments. Also with `web_responsive` module the module's icon is not shown on the apps/root menu.

**Comment by jarroyomorales** [comment]
@Nikul-Chaudhary Do you want me to do a PR to your branch fixing this?

**Comment by norlinhenrik** [comment]
@Nikul-Chaudhary @jarroyomorales Thank you for working on this migration!

File: donation_base/models/product.py
Comment (tbaden) [reply donation_base/models/product.py:69]: it's still v12, api.multi is not set by default here.

File: donation_base/models/product.py
Comment (tbaden) [reply donation_base/models/product.py:76]: it's still v12, api.multi is not set by default here.

File: donation_base/models/partner.py
Comment (tbaden) [reply donation_base/models/partner.py:40]: I think it's because partner.tax_receipt_ids returns False if empty, which has no len() as it's a boolean.

File: donation_base/models/donation_tax_receipt.py
Comment (tbaden) [reply donation_base/models/donation_tax_receipt.py:62]: not mandatory, but would be nice to have it identical in every module:
```
            if values.get('name', '/') == '/':
                seq = self.env['ir.sequence']
                values['name'] = seq.with_context(
                     date=date).next_by_code('donation.tax.receipt') or '/'
```

I would propose to do this after the migration merge, as all other modules for v12 depend on this one.
How is your opinion on this @jarroyomorales ?

**Review by benbrich — APPROVED** [review]
Functionally legit! Thanks for the great work!

**Review by tbaden — APPROVED** [review]
Code Review LGTM
waiting for @jarroyomorales response

File: donation_base/models/partner.py
Comment (jarroyomorales) [reply donation_base/models/partner.py:40]: Isnt it better to just do `partner.tax_receipt_count = len(partner.tax_receipt_ids.ids)` ?

File: donation_base/models/donation_tax_receipt.py
Comment (jarroyomorales) [reply donation_base/models/donation_tax_receipt.py:62]: I think it can be done in the migration itself

File: donation_base/models/partner.py
Comment (tbaden) [reply donation_base/models/partner.py:40]: if partner.tax_receipt_ids is False, than there is not attribute ids, which will result in an error.

File: donation_base/models/donation_tax_receipt.py
Comment (tbaden) [reply donation_base/models/donation_tax_receipt.py:62]: @nikul-serpentcs how would you like to proceed?

Comment (nikul-serpentcs) [reply donation_base/models/donation_tax_receipt.py:62]: > I think it can be done in the migration itself

Yes Sure

File: donation_base/models/partner.py
Comment (jarroyomorales) [reply donation_base/models/partner.py:40]: If I'm not wrong, if partner.tax_receipt_ids is False, partner.tax_receipt_ids.ids returns an empty list [ ].

File: donation_base/models/donation_tax_receipt.py
Comment (tbaden) [reply donation_base/models/donation_tax_receipt.py:62]: @nikul-serpentcs will you do it soon? otherwise I would like to merge this PR so we can proceed with all depending PRs for 12.0 and create an improvement issue for this topic

**Comment by tbaden** [comment]
as stated above I opened an improvement issue: https://github.com/OCA/donation/issues/80
as the migration itself is done. I'll merge it into 12.0

**Pull request merged**
